src/AuthRequired.tsx

```diff
@@ -0,0 +1,108 @@
+ */
+
+import * as React from 'react';
+import { useOktaAuth, OnAuthRequiredFunction } from './OktaContext';
+import * as ReactRouterDom from 'react-router-dom';
+import { toRelativeUrl, AuthSdkError } from '@okta/okta-auth-js';
+// Important! Don't import OktaContext from './OktaContext'
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { OktaContext } from '@okta/okta-react';
+import OktaError from './OktaError';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+let Outlet: any;
+if ('Outlet' in ReactRouterDom) {
+  // trick static analyzer to avoid "'Outlet' is not exported" error
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Outlet = (ReactRouterDom as any)['Outlet' in ReactRouterDom ? 'Outlet' : ''];
+} else {
+  // throw when useMatch is triggered
+  Outlet = () => { 
+    throw new AuthSdkError('Unsupported: AuthRequired only works with react-router-dom v6 or any router library with compatible APIs. See examples under the "samples" folder for how to implement your own custom SecureRoute Component.');
+  };
+}
+
+const AuthRequired: React.FC<{
+  onAuthRequired?: OnAuthRequiredFunction;
+  errorComponent?: React.ComponentType<{ error: Error }>;
+} & React.HTMLAttributes<HTMLDivElement>> = ({ 
+  onAuthRequired,
+  errorComponent,
+  ...props
+}) => {
+  // Need to use OktaContext imported from `@okta/okta-react`
+  // Because AuthRequired needs to be imported from `@okta/okta-react/react-router-6`
+  const { oktaAuth, authState, _onAuthRequired } = useOktaAuth(OktaContext);
+  const pendingLogin = React.useRef(false);
+  const [handleLoginError, setHandleLoginError] = React.useState<Error | null>(null);
+  const ErrorReporter = errorComponent || OktaError;
+
+  React.useEffect(() => {
+    const handleLogin = async () => {
+      if (pendingLogin.current) {
+        return;
+      }
+
+      pendingLogin.current = true;
+
+      const originalUri = toRelativeUrl(window.location.href, window.location.origin);
+      oktaAuth.setOriginalUri(originalUri);
+      const onAuthRequiredFn = onAuthRequired || _onAuthRequired;
+      if (onAuthRequiredFn) {
+        await onAuthRequiredFn(oktaAuth);
+      } else {
+        await oktaAuth.signInWithRedirect();
+      }
+    };
+
+    if (!authState) {
+      return;
+    }
+
+    if (authState.isAuthenticated) {
+      pendingLogin.current = false;
+      return;
+    }
+
+    // Start login if app has decided it is not logged in and there is no pending signin
+    if(!authState.isAuthenticated) { 
+      handleLogin().catch(err => {
+        setHandleLoginError(err as Error);
+      });
+    }  
+
+  }, [
+    authState,
+    oktaAuth,
+    onAuthRequired,
+    _onAuthRequired
+  ]);
+
+  if (handleLoginError) {
+    return <ErrorReporter error={handleLoginError} />;
+  }
+
+  if (!authState || !authState.isAuthenticated) {
```

Comment:
I think it's safer if we flip this boolean

```
if (authState && authState.isAuthenticated) {
    return (
      <Outlet
        { ...props }
      />
  );
  }

  return null
  ```

Reply:
Maybe this should `return <Loader />` rather than null?

Reply:
Added `loadingElement` prop